### PR TITLE
Minimal Worker support

### PR DESCRIPTION
**Pull request by ry**
This adds the ability to spawn additional Isolates from Rust and send
and receive messages from them. This is preliminary work to support
running the typescript compiler in a separate isolate and thus support
native ES modules (#975).

This is a prerequisite commit to #1460

### CODE REVIEW

**Comment by shreeve** [comment]
Does this allow additional isolates such as one for CoffeeScript 2 (which supports native ES modules)?

File: src/resources.rs [hunk 1]
```diff
@@ -111,7 +117,7 @@ pub fn table_entries() -> Vec<(u32, String)> {
 fn test_table_entries() {
   let mut entries = table_entries();
   entries.sort();
-  assert_eq!(entries.len(), 3);
+  // assert_eq!(entries.len(), 3);
```
Comment (ry) [inline src/resources.rs]: Because a worker may be added the number of resources is no longer just the stdio....

File: js/globals.ts [hunk 2]
```diff
@@ -86,3 +87,8 @@ window.TextEncoder = textEncoding.TextEncoder;
 export type TextEncoder = textEncoding.TextEncoder;
 window.TextDecoder = textEncoding.TextDecoder;
 export type TextDecoder = textEncoding.TextDecoder;
+
+window.workerMain = workers.workerMain;
+// TODO These shouldn't be available in main isolate.
```
Comment (piscisaureus) [inline js/globals.ts]: Or should they ... ? ;)

File: js/workers.ts [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+  msg.WorkerGetMessage.startWorkerGetMessage(builder);
+  const inner = msg.WorkerGetMessage.endWorkerGetMessage(builder);
+  const baseRes = await dispatch.sendAsync(
+    builder,
+    msg.Any.WorkerGetMessage,
+    inner
+  );
+  assert(baseRes != null);
+  assert(
+    msg.Any.WorkerGetMessageRes === baseRes!.innerType(),
```
Comment (piscisaureus) [inline js/workers.ts]: Flatbuffers man...

File: src/ops.rs [hunk 4]
```diff
@@ -238,19 +244,19 @@ fn serialize_response(
...
 #[inline]
-fn odd_future(err: DenoError) -> Box<Op> {
+pub fn odd_future(err: DenoError) -> Box<Op> {
   Box::new(futures::future::err(err))
 }
 
 // https://github.com/denoland/deno/blob/golang/os.go#L100-L154
 fn op_code_fetch(
-  state: &IsolateState,
+  state: &Arc<IsolateState>,
```
Comment (piscisaureus) [inline src/ops.rs]: I wonder if we should refactor so that IsolateState itself has Send, instead of wrapping in an Arc everywhere. After all the Mutex is already internalized...

File: src/ops.rs [hunk 5]
```diff
@@ -1530,3 +1536,90 @@ fn op_run_status(
...
+      None => panic!("expected worker_channels"),
+      Some(ref wc) => {
+        let mut wc = wc.lock().unwrap();
+        wc.1.poll()
+      }
+    }
+  }
+}
+
+fn op_worker_get_message(
```
Comment (piscisaureus) [inline src/ops.rs]: Seeing this does make me feel a bit stronger in favor of a unified messaging substrate instead of having distinct ones.

File: src/workers.rs [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+use futures::sync::mpsc;
+use futures::sync::oneshot;
+use futures::Future;
+use std::sync::Arc;
+use std::thread;
+
+/// Rust interface for WebWorkers.
+pub struct Worker {
+  isolate: Isolate,
```
Comment (piscisaureus) [inline src/workers.rs]: Is Isolate itself just a wrapper around a pointer, or a class with lots of data members?
Asking because if the latter is the case, this inflates every entry in the resource table to something potentially huge.

File: src/resources.rs [hunk 7]
```diff
@@ -284,6 +291,54 @@ pub fn add_repl(repl: Repl) -> Resource {
   Resource { rid }
 }
 
+pub fn add_worker(wc: WorkerChannels) -> Resource {
+  let rid = new_rid();
+  let mut tg = RESOURCE_TABLE.lock().unwrap();
+  let r = tg.insert(rid, Repr::Worker(wc));
```
Comment (piscisaureus) [inline src/resources.rs]: Same question as we had once with HTTP -- where does the worker get removed from the resource table?

File: src/ops.rs (hunk 4 above)
Comment (ry) [reply src/ops.rs]: Sure. These types of refactors are generally easy to do due to the type system. I'd leave this as is for now because of laziness...

File: src/ops.rs (hunk 5 above)
Comment (ry) [reply src/ops.rs]: We do have a unified messaging substrate, `getMessage` and `postMessage` are built on top of the lower-level op dispatch system.

I'd rephrase your suggestion: "`postMessage` should be the means that we dispatch ops."

I'm not so convinced of that. I think we have very particular concerns in the IPC system that the WebWorkers API doesn't address like synchronous responses, zero-copy buffers.

File: src/workers.rs (hunk 6 above)
Comment (ry) [reply src/workers.rs]: Yes, it's a class with data members - it's not just a pointer. Your comment is valid.

I agree that the resource table should have `Box` elements instead of directly containing structs - but I'd rather do that refactor in a different PR.

File: src/resources.rs (hunk 7 above)
Comment (ry) [reply src/resources.rs]: They weren't being correctly removed. I've fixed this now. When the thread exits it calls `resource.close()` which removes it from the table. I've added a test to ensure this.

File: src/resources.rs (hunk 1 above)
Comment (ry) [reply src/resources.rs]: No longer commented out.
